### PR TITLE
[XPU][TritonGEN] Replace subgroup reduce and scan

### LINKED ISSUE #2892: Replace SubGroupReduce and SubGroupScan operations from TritonGEN dialect with SPIR-V dialect operations
Similar to #2815, replace TritonGEN dialect operations SubgroupReduce and SubgroupScan with equivalent operations from the upstream SPIR-V dialect.

### PR DESCRIPTION
**Pull request by sommerlukas**
Replace usage of the TritonGEN dialect subgroup reduce and scan operations with equivalent operations from the SPIR-V dialect.

This closes #2892.

### CODE REVIEW

File: third_party/intel/lib/TritonIntelGPUToLLVM/TargetInfo.cpp [hunk 1]
```diff
@@ -107,6 +108,77 @@ Value TargetInfo::programId(RewriterBase &rewriter, Location loc,
...
+                rewriter, loc, resultType, acc, numLanesToReduce, warpSize);
+          })
+          .Case<arith::XOrIOp>([&](auto) {
+            return createSPIRVGroupOp<spirv::GroupNonUniformBitwiseXorOp>(
+                rewriter, loc, resultType, acc, numLanesToReduce, warpSize);
+          })
+          .Default([](auto) {
+            llvm_unreachable("Unsupported reduction");
+            return Value();
+          });
```
Comment (victor-eds) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TargetInfo.cpp:176]:
```suggestion
template <typename OpTy>
struct spirv_group_op {};

template <>
struct spirv_group_op<arith::AddFOp> {
  using type = spirv::GroupNonUniformFAddOp;
};
// ...
template <typename OpTy>
using spirv_group_op_ty = spirv_group_op<OpTy>::type;
// ----
  Value warpReduce =
      TypeSwitch<mlir::Operation *, Value>(reduceOp)
          .Case<arith::AddFOp, arith::AddIOp, arith::MulFOp, arith::MulIOp, arith::MaxNumFOp, arith::MinNumFOp, arith::AndIOp, arith::OrIOp, arith::XOrIOp>([&](auto reduceOp) {
            return createSPIRVGroupOp<spirv_group_op_ty<decltype(reduceOp)>>(
                rewriter, loc, resultType, acc, numLanesToReduce, warpSize);
          });
```
No need for that default as this is implicit.

Struct type approach looks cleaner to me.

`TypeSwitch` is defined under `::mlir` as an alias

File: third_party/intel/lib/TritonIntelGPUToLLVM/TargetInfo.cpp [hunk 2]
```diff
@@ -134,28 +206,22 @@ bool TargetInfo::warpReduce(RewriterBase &rewriter, Location loc,
...
-          .Case<arith::OrIOp>([&](auto) { return TritonGEN::ReduceKind::OR; })
-          .Case<arith::XOrIOp>([&](auto) { return TritonGEN::ReduceKind::XOR; })
-          .Default([](auto) { return std::nullopt; });
-  if (reduceKind == std::nullopt)
+  auto supportedOp =
+      llvm::TypeSwitch<mlir::Operation *, bool>(reduceOp)
+          .Case<arith::AddFOp, arith::AddIOp, arith::MulFOp, arith::MulIOp,
+                arith::MaxNumFOp, arith::MinNumFOp, arith::AndIOp, arith::OrIOp,
+                arith::XOrIOp>([&](auto) { return true; })
+          .Default([](auto) { return false; });
```
Comment (victor-eds) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TargetInfo.cpp:214]:
```suggestion
  auto supportedOp = isa<arith::AddFOp, arith::AddIOp, arith::MulFOp, arith::MulIOp,
                arith::MaxNumFOp, arith::MinNumFOp, arith::AndIOp, arith::OrIOp,
                arith::XOrIOp>(reduceOp);
```

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp [hunk 3]
```diff
@@ -585,17 +586,25 @@ class ReduceOpConversion : public ConvertTritonGPUOpToLLVMPattern<ReduceOp> {
     using AllReduceOperation = mlir::gpu::AllReduceOperation;
     AllReduceOperation redKind;
     if (isa<arith::AddFOp>(combine))
```
Comment (victor-eds) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp:588]:
```cpp
TypeSwitch<Operation *>(combine)
    .Case<arith::AddFOp, arith::MaxNumFOp>([&](auto combine) {
      rewriter.replaceOpWithNewOp<spirv_group_op_ty<decltype(combine)>>(
        combine, typeConverter->convertType(combine.getType(0)), spirv::Scope::Subgroup,
        spirv::GroupOperation::Reduce, adaptor.getSrcs()[0], Value());
    });
```

This or similar. Also, no `else` needed as that's covered by `TypeSwitch` semantics.

Feel free to adapt by requirements or likings.

Would drop aux. function.

**Review by victor-eds — COMMENTED** [review]
Looking way better. Just some stylistic comments. Feel free to apply at your liking. LGTM otherwise.

**Comment by sommerlukas** [comment]
> Please fix `test_reduce[...xor_sum...]` failures.

Yeah, I already have a hunch what's causing it, will fix 👍

**Review by etiotto — APPROVED** [review]

**Review by whitneywhtsang — APPROVED** [review]

**Pull request merged**